### PR TITLE
Introduce `DTypePolicyMap`

### PR DESCRIPTION
**Pull request by james77777778**
ORIGINAL FROM #19783
Sorry for the inconvenience.

I have tried to implement the idea mentioned by @mattdangerw :
https://github.com/keras-team/keras-nlp/pull/1612

I think this idea is clean and only requires some small changes to support mixing dtype policies into subclasses.

In general, `DTypePolicyMap` is similar to `LayoutMap` as it contains a dict to record the (layer name, policy) pair.
The limitations / overheads of using `DTypePolicyMap`:
- Each layer must have a unique name (`auto_name` might fail)
- Some extra logic needs to be added in `get_config` for the top class (ex: `Backbone` in KerasCV and KerasNLP)

The basic usage  (adding quantization support for `MultiHeadAttention`):

<details>

```python
import keras
from keras import dtype_policies
from keras import ops
import numpy as np

@keras.saving.register_keras_serializable("MyPackage")
class MyMultiHeadAttention(keras.layers.MultiHeadAttention):
    def get_config(self):
        config = super().get_config()
        # `default_policy=None` enables us to defer to
        # `keras.config.dtype_policy` during loading.
        dtype_policy_map = dtype_policies.DTypePolicyMap(default_policy=None)
        for layer in self._flatten_layers():
            if layer.dtype_policy.is_quantized:
                dtype_policy_map[layer.name] = layer.dtype_policy
        if len(dtype_policy_map) > 0:
            config.update({"dtype": dtype_policy_map})
        return config


query = keras.Input(shape=[4, 6])
value = keras.Input(shape=[4, 6])
outputs = MyMultiHeadAttention(2, 4, name="my_mha")(query, value)
model = keras.Model([query, value], outputs)

x_query = keras.random.uniform([2, 4, 6])
x_value = keras.random.uniform([2, 4, 6])
print(f"Saving stage, global policy: {keras.config.dtype_policy()}")
model.quantize("int8")
for layer in model._flatten_layers(include_self=False, recursive=True):
    print(layer.name, layer.dtype_policy)
model.save("model.keras")
y = model([x_query, x_value])

keras.config.set_dtype_policy("mixed_bfloat16")
print(f"Loading stage, global policy: {keras.config.dtype_policy()}")
new_model = keras.models.load_model("model.keras")
for layer in new_model._flatten_layers(include_self=False, recursive=True):
    print(layer.name, layer.dtype_policy)
ref_y = new_model([x_query, x_value])

np.testing.assert_allclose(
    ops.convert_to_numpy(y),
    ops.convert_to_numpy(ops.cast(ref_y, "float32")),
    atol=1e-2,
)

```

</details>

Outputs:
```bash
Saving stage, global policy: <FloatDTypePolicy "float32">
input_layer <FloatDTypePolicy "float32">
input_layer_1 <FloatDTypePolicy "float32">
my_mha <FloatDTypePolicy "float32">
attention_output <QuantizedDTypePolicy "int8_from_float32">
dropout <FloatDTypePolicy "float32">
softmax <FloatDTypePolicy "float32">
value <QuantizedDTypePolicy "int8_from_float32">
key <QuantizedDTypePolicy "int8_from_float32">
query <QuantizedDTypePolicy "int8_from_float32">

Loading stage, global policy: <FloatDTypePolicy "bfloat16">
input_layer <FloatDTypePolicy "mixed_bfloat16">
input_layer_1 <FloatDTypePolicy "mixed_bfloat16">
my_mha <FloatDTypePolicy "mixed_bfloat16">
attention_output <QuantizedDTypePolicy "int8_from_mixed_bfloat16">
dropout_1 <FloatDTypePolicy "mixed_bfloat16">
softmax_1 <FloatDTypePolicy "mixed_bfloat16">
value <QuantizedDTypePolicy "int8_from_mixed_bfloat16">
key <QuantizedDTypePolicy "int8_from_mixed_bfloat16">
query <QuantizedDTypePolicy "int8_from_mixed_bfloat16">
```

Some updates:
- A high-level description has been added, as suggested by @mattdangerw 
- An internal property, `dtype_argument`, has been introduced to store the passed argument. `dtype_argument` can be further utilized in `build` (e.g., in `MultiHeadAttention`).

### CODE REVIEW

**Comment by james77777778** [comment]
Hi @mattdangerw @fchollet 

I have successfully brought quantization support for Gemma with this PR.
This should provide high confidence that `DTypePolicyMap` works.

Here is the implementation in just 1 script:
(The actual modifications are small)

<details>

```python
import numpy as np
from keras_nlp.src.layers.modeling.reversible_embedding import (
    ReversibleEmbedding,
)
from keras_nlp.src.layers.modeling.rotary_embedding import RotaryEmbedding
from keras_nlp.src.models.backbone import Backbone
from keras_nlp.src.models.gemma.gemma_attention import CachedGemmaAttention
from keras_nlp.src.models.gemma.gemma_backbone import GemmaBackbone
from keras_nlp.src.models.gemma.gemma_decoder_block import GemmaDecoderBlock
from keras_nlp.src.models.gemma.rms_normalization import RMSNormalization

import keras
from keras import dtype_policies
from keras import ops


@keras.saving.register_keras_serializable()
class MyCachedGemmaAttention(CachedGemmaAttention):
    def build(self, inputs_shape):
        self.hidden_dim = inputs_shape[-1]
        self.query_dense = keras.layers.EinsumDense(
            "btd,ndh->btnh",
            output_shape=(None, self.num_query_heads, self.head_dim),
            kernel_initializer=self._kernel_initializer,
            dtype=self.dtype_argument,
            name="query",
        )
        self.query_dense.build(inputs_shape)
        self.key_dense = keras.layers.EinsumDense(
            "bsd,kdh->bskh",
            output_shape=(None, self.num_key_value_heads, self.head_dim),
            kernel_initializer=self._kernel_initializer,
            dtype=self.dtype_argument,
            name="key",
        )
        self.key_dense.build(inputs_shape)
        self.value_dense = keras.layers.EinsumDense(
            "bsd,kdh->bskh",
            output_shape=(None, self.num_key_value_heads, self.head_dim),
            kernel_initializer=self._kernel_initializer,
            dtype=self.dtype_argument,
            name="value",
        )
        self.value_dense.build(inputs_shape)
        self.dropout_layer = keras.layers.Dropout(
            rate=self.dropout,
            dtype=self.dtype_argument,
        )
        self.output_dense = keras.layers.EinsumDense(
            equation="btnh,nhd->btd",
            output_shape=(None, self.hidden_dim),
            kernel_initializer=self._kernel_initializer,
            dtype=self.dtype_argument,
            name="attention_output",
        )
        self.output_dense.build(
            (None, None, self.num_query_heads, self.head_dim)
        )
        self.softmax = keras.layers.Softmax(dtype="float32")
        self.rope_layer = RotaryEmbedding(
            max_wavelength=10_000.0, dtype=self.dtype_argument
        )
        self.built = True


@keras.saving.register_keras_serializable()
class MyGemmaDecoderBlock(GemmaDecoderBlock):
    def __init__(
        self,
        hidden_dim,
        intermediate_dim,
        head_dim,
        num_query_heads,
        num_key_value_heads,
        layer_norm_epsilon=1e-6,
        dropout=0,
        **kwargs,
    ):
        keras.layers.Layer.__init__(self, **kwargs)
        self.intermediate_dim = intermediate_dim
        self.hidden_dim = hidden_dim
        self.num_query_heads = num_query_heads
        self.num_key_value_heads = num_key_value_heads
        self.head_dim = head_dim
        self.layer_norm_epsilon = layer_norm_epsilon
        self.dropout = dropout
        self.pre_attention_norm = RMSNormalization(
            epsilon=self.layer_norm_epsilon,
            dtype=self.dtype_argument,
            name="pre_attention_norm",
        )
        self.attention = MyCachedGemmaAttention(
            head_dim=head_dim,
            num_query_heads=num_query_heads,
            num_key_value_heads=num_key_value_heads,
            dropout=dropout,
            dtype=self.dtype_argument,
            name="attention",
        )
        if self.dropout > 0:
            self.attention_dropout = keras.layers.Dropout(rate=dropout)
            self.feedforward_dropout = keras.layers.Dropout(rate=dropout)
        self.pre_ffw_norm = RMSNormalization(
            epsilon=self.layer_norm_epsilon,
            dtype=self.dtype_argument,
            name="pre_ffw_norm",
        )
        self.gating_ffw = keras.layers.EinsumDense(
            equation="btd,df->btf",
            output_shape=(None, self.intermediate_dim // 2),
            dtype=self.dtype_argument,
            name="ffw_gating",
        )
        self.gating_ffw_2 = keras.layers.EinsumDense(
            equation="btd,df->btf",
            output_shape=(None, self.intermediate_dim // 2),
            dtype=self.dtype_argument,
            name="ffw_gating_2",
        )
        self.ffw_linear = keras.layers.EinsumDense(
            equation="btf,fd->btd",
            output_shape=(None, self.hidden_dim),
            dtype=self.dtype_argument,
            name="ffw_linear",
        )


@keras.saving.register_keras_serializable()
class MyGemmaBackbone(GemmaBackbone):
    def __init__(
        self,
        vocabulary_size,
        num_layers,
        num_query_heads,
        num_key_value_heads,
        hidden_dim,
        intermediate_dim,
        head_dim,
        layer_norm_epsilon=1e-6,
        dropout=0,
        dtype=None,
        dtype_policy_map=None,
        **kwargs,
    ):
        if dtype_policy_map is not None:
            dtype_policy_map = keras.dtype_policies.deserialize(
                dtype_policy_map
            )
        if dtype is None and dtype_policy_map is not None:
            dtype_policy_map._default_policy = keras.dtype_policies.get(dtype)
            dtype = dtype_policy_map
        self.token_embedding = ReversibleEmbedding(
            input_dim=vocabulary_size,
            output_dim=hidden_dim,
            tie_weights=True,
            embeddings_initializer=keras.initializers.VarianceScaling(
                scale=1.0,
                mode="fan_in",
                distribution="untruncated_normal",
                seed=None,
            ),
            dtype=dtype,
            name="token_embedding",
        )
        self.transformer_layers = []
        for i in range(num_layers):
            layer = MyGemmaDecoderBlock(
                intermediate_dim=intermediate_dim,
                hidden_dim=hidden_dim,
                num_query_heads=num_query_heads,
                head_dim=head_dim,
                num_key_value_heads=num_key_value_heads,
                dropout=dropout,
                dtype=dtype,
                name=f"decoder_block_{i}",
            )
            self.transformer_layers.append(layer)
        self.layer_norm = RMSNormalization(
            epsilon=layer_norm_epsilon,
            dtype=dtype,
            name="final_normalization",
        )
        # === Functional Model ===
        token_id_input = keras.Input(
            shape=(None,), dtype="float32", name="token_ids"
        )
        padding_mask_input = keras.Input(
            shape=(None,), dtype="float32", name="padding_mask"
        )
        x = self.token_embedding(token_id_input)
        x = x * ops.cast(ops.sqrt(hidden_dim), x.dtype)
        for transformer_layer in self.transformer_layers:
            x = transformer_layer(x, padding_mask=padding_mask_input)
        sequence_output = self.layer_norm(x)
        Backbone.__init__(
            self,
            inputs={
                "token_ids": token_id_input,
                "padding_mask": padding_mask_input,
            },
            outputs=sequence_output,
            dtype=(
                dtype
                if not isinstance(dtype, keras.dtype_policies.DTypePolicyMap)
                else dtype.default_policy
            ),
            **kwargs,
        )
        # === Config ===
        self.vocabulary_size = vocabulary_size
        self.num_layers = num_layers
        self.num_query_heads = num_query_heads
        self.num_key_value_heads = num_key_value_heads
        self.hidden_dim = hidden_dim
        self.intermediate_dim = intermediate_dim
        self.head_dim = head_dim
        self.layer_norm_epsilon = layer_norm_epsilon
        self.dropout = dropout

    def get_config(self):
        config = super().get_config()
        dtype_policy_map = dtype_policies.DTypePolicyMap(default_policy=None)
        for layer in self._flatten_layers():
            if (
                layer.dtype_policy.is_quantized
                and layer.name not in dtype_policy_map
            ):
                dtype_policy_map[layer.name] = layer.dtype_policy
        if len(dtype_policy_map) > 0:
            config.update({"dtype_policy_map": dtype_policy_map})
        return config


gemma_backbone = MyGemmaBackbone(256128, 2, 4, 4, 128, 256, 128, 1e-6)
token_ids = keras.ops.ones((2, 5), "int32")
padding_mask = keras.ops.ones((2, 5), "int32")
inputs = {"token_ids": token_ids, "padding_mask": padding_mask}

gemma_backbone.quantize("int8")
print("`gemma_backbone` has been quantized to int8")
y = gemma_backbone(inputs)
config = gemma_backbone.get_config()
gemma_backbone.save("model.keras")

reloaded_gemma_backbone = keras.saving.load_model("model.keras")
reloaded_y = reloaded_gemma_backbone(inputs)

np.testing.assert_allclose(
    keras.ops.convert_to_numpy(y),
    keras.ops.convert_to_numpy(reloaded_y),
)
print("The outputs are all closed!")

```

</details>

The modifications are as follows:
- Replace the plain `dtype` with `dtype_argument` introduced by this PR.
`dtype_argument` is the passed `dtype` argument store in `Operation` and it could be a str, `DTypePolicy` or `DTypePolicyMap`
- Determine the `dtype` passed to subclasses and sublayers in `GemmaBackbone.__init__`

That's all. Super simple. It should even be feasible to do this in `Backbone` and apply it to all models.

Please let me know what you think. Thanks!

**Comment by fchollet** [comment]
Looking good -- I'll wait to hear from @mattdangerw before merging.

**Comment by mattdangerw** [comment]
@james77777778 looks good!

Still a couple things that I think we might want to fix up with the high level UX.
1) Layer names are often not unique.
2) The pathing for this and `keras.distribution.LayoutMap` are very different. Which feels confusing for users.

I'm gonna put together a quick commit on how we could deal with this and share it with you, just to get some thoughts. Or I'll try to do it today, and realize a bunch of places I'm missing things :). Either way I'll post back here soon!

**Comment by mattdangerw** [comment]
Ok, I'll keep testing this out, but just to share early the idea is here... 

https://github.com/james77777778/keras/compare/dtype-policy-map-2...mattdangerw:keras:dtype-policy-map-idea

- Make the map another `DTypePolicy` subclass.
- Add a `self.quantization_mode` property on the base layer, much like `self.compute_dtype` and `self.variable_dtype`. 
- Have all those properties on the base layer pull the correct policy form the map.
- Lastly, use the build time name scope to look up by keys. So the names work the same way as `keras.distribution.LayoutMap`.

**Comment by mattdangerw** [comment]
Here's a colab with a little more usage -> https://colab.research.google.com/gist/mattdangerw/aee1ce80eb8128dff3911ea3442d68af/dtypemap-usages.ipynb

I think there's some advantages here.
- Everything passed as dtype resolves to a `DTypePolicy` object.
- Supports global and local setting of these policy maps.
- Supports specifying policies by the full name scope paths, e.g. `"encoder/.*/query"` and `decoder/.*/value`.
- Does not change the way dtype should be proxied from layer -> sublayer. Pass `self.dtype_policy`.
- Delegating resolving `quantization_mode` to the base layer is similar to how we handle `compute_dtype` and `variable_dtype`. Simpler division of responsibilities, cleans up the dense/embedding layers a bit.

Very possible I am still missing some cases though. Let me know what you think!

**Comment by james77777778** [comment]
Hey @mattdangerw 
This is great. It really simplifies the high-level logic!

The only issue I see is that we now rely on `self.built` to get the correct policy.
The properties of the layer might be wrong before the `build`, no?

Could I merge these changes and see how it goes, or do you prefer building it in your fork?

**Comment by mattdangerw** [comment]
> Could I merge these changes and see how it goes, or do you prefer building it in your fork?

Yeah, merge away!

Re the build before/after problem. Yeah, I couldn't really figure out a way around that. Basically, we don't have a reliable way to get the name scope until we have built. We could consider extend the name scope to layer init time, but layers might be initialized in a weird order (e.g. function model is layers first, than model). It's really only at build/call time we can construct the correct nested name scope.

I was content to let that be a quirk of how these are applied, but @fchollet do you have any thoughts here?

**Comment by james77777778** [comment]
I have fixed the CI and made some minor changes:

- Simplify the logic of `*_dtype` by utilizing `self.dtype_policy`
- Rely on `self.build_name_scope` instead of `self.built` in `self.dtype_policy` because we need the correct policy during `build`
- Raise an error if there is a nested `DTypePolicyMap`
- Fix the `dtype_policy` setter
- Fix the `quantize` in `Dense`, `Einsum` and `Embedding`

Should be ready for reviewing.
@mattdangerw @fchollet

File: keras/src/dtype_policies/dtype_policy_map.py
Comment (fchollet) [inline keras/src/dtype_policies/dtype_policy_map.py:11]: "strings" -> layer paths

To save space you can remove "A " and switch "that maps" to "mapping"

File: keras/src/dtype_policies/dtype_policy_map.py
Comment (fchollet) [inline keras/src/dtype_policies/dtype_policy_map.py:42]: "In the normal case, the key to query is usually the `layer.name`" -> this is slightly confusing. In what cases would this *not* be true?

File: keras/src/dtype_policies/dtype_policy_map.py
Comment (fchollet) [inline keras/src/dtype_policies/dtype_policy_map.py:47]: Should we introduce a `layer.path` property similar to `variable.path`?

File: keras/src/dtype_policies/dtype_policy_map.py
Comment (james77777778) [reply keras/src/dtype_policies/dtype_policy_map.py]: Fixed

File: keras/src/dtype_policies/dtype_policy_map.py
Comment (james77777778) [reply keras/src/dtype_policies/dtype_policy_map.py]: I have modified the docstring here. Should be clear now.

File: keras/src/dtype_policies/dtype_policy_map.py
Comment (james77777778) [reply keras/src/dtype_policies/dtype_policy_map.py:44]: Introducing `Layer.path` makes it much cleaner. Done.

**Comment by fchollet** [comment]
LGTM -- will defer to @mattdangerw for final approval. Thank you!

File: keras/src/layers/layer.py
Comment (mattdangerw) [inline keras/src/layers/layer.py:693]: Doesn't this break things like `MultiHeadAttention`? In the changes I was playing with, we only resolved `path` in the `compute_dtype`, `variable_dtype`, and `quatization_mode` properties. That was so that composite layer could just pass down the dtype policy to children. Like this

https://github.com/keras-team/keras/blob/2f846b0cfeafbe9004e727c571f0e496057201b1/keras/src/layers/attention/multi_head_attention.py#L295

If we want `dtype=self.dtype_policy` to be the way to propagate dtype, I think we need to flip back to how this was--only resolve path in `compute_dtype`, `variable_dtype`, and `quatization_mode` getters.

File: keras/src/layers/core/embedding.py
Comment (mattdangerw) [inline keras/src/layers/core/embedding.py:382]: Can we actually make this go through the layer `dtype_policy` setter?

```python
policy = dtype_policies.get(f"{mode}_from_{self.dtype_policy.name}")
self.dtype_policy = policy
```

I would be really great if layer implementations (dense/embedding/etc) had no idea about `DTypePolicyMap` class, and the map stays a superclass implementation detail. That would allow people to write much simple layer subclasses.

Seems like we could do that, we add an early return at the top of this method?

```python
if self._is_quantized:
    return
```

Maybe there's a better way to do this, but any way we could get the next 6 lines to an implementation details of the base layer that would probably be a good thing.

File: keras/src/layers/layer.py
Comment (james77777778) [reply keras/src/layers/layer.py]: Thanks! I wasn't aware of that.
I have added a test in MHA to prevent future breakage.

File: keras/src/layers/core/embedding.py
Comment (james77777778) [reply keras/src/layers/core/embedding.py:380]: I agree with you. It is feasible, and I have made the update.

**Review by mattdangerw — APPROVED** [review]
Great work! This LGTM!

**Comment by james77777778** [comment]
The failure seems to be an unrelated issue.
I was able to successfully pass all tests with my CUDA rig.

One question:
Once this PR is merged, will there be a new release of Keras?
If so, I can proceed to implement quantization support for KerasNLP.

**Comment by fchollet** [comment]
> Once this PR is merged, will there be a new release of Keras?
> If so, I can proceed to implement quantization support for KerasNLP.

Awesome -- yes, we'll release a new Keras version on Monday.

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution!

**Pull request merged**
